src/tools/diagnostics/analyze_resource/analyze_resource.js

```diff
@@ -0,0 +1,166 @@
+        process.exit(1);
+    }
+    process.exit(0);
+}
+
+function check_supported_resource_type(resource_type) {
+    const SUPPORTED_TYPES_OF_RESOURCES = ['aws-s3', 's3-compatible', 'ibm-cos', 'google-cloud-storage', 'azure-blob']; //'pv-pool' and 'nsfs' are not cloud vendor and we will not check them
+    if (SUPPORTED_TYPES_OF_RESOURCES.includes(resource_type)) {
+        console.info('Diagnose Resource Started');
+    } else {
+        throw new Error(`${resource_type} is not supported`);
+    }
+}
+
+async function test_resource(cloud_vendor, bucket) {
+    check_arguments(cloud_vendor, bucket);
+    console.info(`We will have several tests to check the status of the resource and permission of the bucket ${bucket}`);
+    const key_to_head = await list_objects_and_get_one_key(cloud_vendor, bucket);
+    let head_object_executed;
+    let head_skip_reason;
+    if (key_to_head) {
+        console.info(`A returned key: ${key_to_head}`);
+        await head_object(cloud_vendor, bucket, key_to_head);
+        head_object_executed = true;
+    } else {
+        head_skip_reason = 'head object after list objects was not executed since there was no returned key';
+    }
+    const key_to_write = `test_resource_${new Date().getTime()}_(can_be_deleted).txt`;
+    const is_write_object = await write_object(cloud_vendor, bucket, key_to_write);
+
+    // in case we wrote an object but we didn't have a returned key from the bucket
+    // we will try to head the written object
+    if (is_write_object && !key_to_head) {
+        await head_object(cloud_vendor, bucket, key_to_write);
+        head_object_executed = true;
+    }
+    if (!is_write_object && !key_to_head) {
+        head_skip_reason = `head object (after write object) skipped since ${key_to_write} was not written to the bucket`;
+    }
+    if (!head_object_executed) {
+        testing_status.skip.push('head object');
+        console.warn(`Test head object in bucket ${bucket} skipped, since ${head_skip_reason}`);
+    }
+
+    // in case we wrote an object we want to delete it
+    await delete_object_after_write(cloud_vendor, bucket, key_to_write, is_write_object);
+    print_test_summary();
+}
+
+function check_arguments(cloud_vendor, bucket) {
+    const test = 'test_resource';
+    if (!cloud_vendor) {
+        console.warn(`${test}: received cloud vendor ${cloud_vendor}`);
+        throw new Error(`Received cloud vendor ${cloud_vendor}. Please provide connection details`);
+    }
+    if (!bucket) {
+        console.warn(`${test}: received bucket ${bucket}`);
+        throw new Error(`Received bucket ${bucket}. Please provide bucket`);
+    }
+}
+
+async function list_objects_and_get_one_key(cloud_vendor, bucket) {
+    const test = 'list objects';
+    try {
+        await cloud_vendor.list_objects(bucket);
+        console.info(`Test ${test} in bucket ${bucket} passed `);
+        testing_status.pass.push(test);
+        return await cloud_vendor.get_key();
```

Comment:
I want to return a value only if the function is completed successfully.
In this case the return statement is in the try block.
What you are suggesting is that the function will always return a value (and then I will return a null), and it's different.

Reply:
It could be okay from functionality point of view and may be in JS it would not cause any issue.
However, it is a good practice to return a value in all cases or not at all, irrespective of any programming language.
Definition of a function is that it take some argument , does some processing and return some value.
Now, a function might not return a value all together (just print a hello world) and that is okay but it is not the right thing that it will return a value in one case but does not return anything in other case.

Reply:
I'll change it.